src/oauth/index.ts

```diff
@@ -0,0 +1,203 @@
+ */
+async function putParam(paramRequest:AWS.SSM.Types.PutParameterRequest): Promise<void> {
+  const putReq = ssmClient.putParameter(paramRequest);
+  try {
+    await putReq.promise();
+    console.debug('successfully updated parameter');
+  } catch (err) {
+    console.warn(err);
+  }
+}
+
+/**
+ * async function responsible for retrieving and store
+ * valid OAuth2 client credentials access tokens into the AWS Parameter Store.
+ *
+ * @param {string} oauthKey Oauth Key as configured via environment variable
+ * @return {string} valid access token based on configured OAuth server
+ */
+async function getAndSaveToken(oauthKey: string): Promise<string> {
+  const secrets = await getParam({
+    Name: `/oauth/${oauthKey}/${auth.CC}/secrets`,
+    WithDecryption: true,
+  });
+
+  if (!secrets) {
+    throw new Error(
+      `no secrets in Parameter Store defined for key ${oauthKey}, type ${auth.CC}`);
+  }
+
+  const parsedSecrets = JSON.parse(secrets);
+
+  const buff:Buffer = Buffer.from(
+    `${parsedSecrets.clientId}:${parsedSecrets.clientSecret}`);
+  const base64Creds = buff.toString('base64');
+
+  const expirationDate = new Date();
+
+  const tokenResp = await p({
+    url: parsedSecrets.endpoint,
+    parse: 'json',
+    method: 'POST',
+    headers: {
+      'Authorization': `Basic ${base64Creds}`,
+    },
+    form: {
+      'grant_type': auth.CC,
+    },
+  });
+
+  const tokenRespBody:any = tokenResp.body;
+
+  if (!tokenRespBody || !tokenRespBody['access_token'] ||
+      !tokenRespBody['expires_in']) {
+    throw new Error('Invalid token response');
+  }
+
+  // set token expiration to be 2 minutes before stated expiration
+  expirationDate.setSeconds(
+    expirationDate.getSeconds() + tokenRespBody['expires_in'] - 120);
+
+  const tokenVal = {
+    token: tokenRespBody['access_token'],
+    expiration: expirationDate.getTime(),
+  };
+
+  const tokenSsmParams:AWS.SSM.Types.PutParameterRequest = {
+    Name: `/oauth/${oauthKey}/${auth.CC}/token`,
+    Type: 'SecureString',
+    Value: JSON.stringify(tokenVal),
+    Overwrite: true,
+  };
+
+  await putParam(tokenSsmParams);
+  return tokenVal.token;
+}
+
+/**
+ * Generates a JWT token based on the default expiration of
+ * 5 minutes and passed api key + secret
+ *
+ * @param {string} apiKey JWT API Key
+ * @param {string} apiSecret JWT API secret
+ * @return {string} valid JWT token
+ */
+function generateJWTToken(apiKey: string, apiSecret: string): string {
+  const expiryTime = new Date().getTime() + DEFAULT_EXPIRY_MILLIS;
+  const payload = {
+    iss: apiKey,
+    exp: expiryTime
+  }
+  return jwtLib.sign(payload, apiSecret);
+}
+
+function checkKey(oauthKey: string) {
+  if (!oauthKey || oauthKey.length === 0) {
+    throw new Error('please pass a valid OAuth key');
+  }
+
+  console.debug('retrieving oauth credentials for:', oauthKey);
+}
+
+type AsyncFunction = () => Promise<String>
+
+export function jwt(oauthKey: string): AsyncFunction {
```

Comment:
if there aren't comments here, do you still get the tooltip highlights on IDEs that render the jsdoc? i do like when they do that

Reply:
oh...no tooltip on hover yeah. I like it more too though! I'll move the comment I already wrote over the jwt/clientCredentials functions.